src/browser/deck/actions.js

```diff
@@ -6,119 +6,218 @@ define((require, exports, module) => {
-    const isCLosingSelected = closing == selected;
-    const isLast = isCLosingSelected && closing == items.count() - 1;
-    const reselected = !isCLosingSelected ? items :
-                       isLast ? selectPrevious(items) :
-                       selectNext(items);
-    return reselected.remove(closing);
-  };
+  const asSelected = item => item.update('isSelected', True);
+  const asActive = item => item.update('isActive', True);
 
-  // Returns index of the last item from the given `items`.
-  const isLast = (item, items) => item.equals(items.last());
+  const asUnselected = item => item.update('isSelected', False);
+  const asInactive = item => item.update('isActive', False);
 
-  // Inesrts item after first item for which `shouldFollow(item)` is true,
-  // if `shouldFollow` is ommited inserts item after last item.
-  const insertAfter = (items, item, shouldFollow=isLast) => items => {
-    const after = items.findIndex(item => shouldFollow(item, items));
-    return items.slice(0, after).push(item).concat(items.slice(after));
-  };
+  const switchSelected = transition(asUnselected, asSelected);
+  const switchActive = transition(asInactive, asActive);
 
-  const append = (items, item) => items.push(item);
+  const select = (items, item) =>
+    switchSelected(items, indexOfSelected(items), items.indexOf(item));
 
-  const isFirst = (item, items) => item.equals(items.first());
+  // Takes `items` and selects item previous to currently selected one.
+  // If selected item is first item, then last item is selected.
+  const selectPrevious = advance(switchSelected,
+                                 indexOfSelected,
+                                 indexOfPrevious);
+  const selectNext = advance(switchSelected,
+                             indexOfSelected,
+                             indexOfNext);
+
+  // Takes deck `items` and activates selected item.
+  const activate = items => {
+    const from = indexOfActive(items);
+    const to = indexOfSelected(items);
+
+    return switchActive(items, from, to);
+  }
 
-  // Inesrts item before first item for which `shouldFollow(item)` is true,
-  // if `shouldFollow` is ommited inserts item before first item.
-  const insertBefore = (items, item, shouldLead=isFirst) => items => {
-    const before = items.findIndex(item => shouldLead(item, items));
-    return before == 0 ? items.unshift(item) :
-           items.slice(0, before - 1).push(item).concat(items.slice(before - 1));
+  const activateNext = compose(activate, selectNext);
+  const activatePrevious = compose(activate, selectPrevious);
+
+  // Take an `items` and optional `p` predicate function and removes
+  // first item where `p(item)` is true. If item to be removed happens
+  // to be selected or active then it takes care of activating / selecting
+  // different item according to logic explained in the inline comments
+  // below.
+  const remove = (items, p=exports.isActive) => {
```

Comment:
> Nit: since isActive is a const in scope, it would be safer to reference isActive directly.

That's what I had originally but it turns out that `isActive` declaration in the function body makes such a function illegal, does not makes sense if you ask me, but...